packages/driver/src/httpScram.ts

```diff
@@ -41,27 +66,42 @@ export async function HTTPSCRAMAuth(
       Authorization: `SCRAM-SHA-256 data=${utf8ToB64(clientFirst)}`,
     },
   });
-  if (serverFirstRes.status === 403) {
-    // tslint:disable-next-line:no-console
-    console.log(serverFirstRes);
-    throw new Error(`Server doesn't support HTTP SCRAM authentication`);
+
+  // The first request must have status 401 Unauthorized and provide a
+  // WWW-Authenticate header with a SCRAM-SHA-256 challenge.
+  // See: https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L153-L157
+  const authenticateHeader = serverFirstRes.headers.get("WWW-Authenticate");
+  if (serverFirstRes.status !== 401 || !authenticateHeader) {
+    const body = await serverFirstRes.text();
+    throw new ProtocolError(`authentication failed: ${body}`);
```

Comment:
I'm not sure I think the body text is the only (or indeed the most helpful) thing we can log here. And in fact, I'm concerned about leaking data to logs or having to even parse the body in these error cases.

I think it'd be helpful to think about some structured way to represent the failed response that has a relatively static size, or at least a reasonably finite size. Something along these lines:

```suggestion
    const headerKeys = Array.from(serverFirstRes.headers().keys());
    const status = serverFirstRes.status;
    throw new ProtocolError(`Authentication failed with status ${status}. Headers: ${headerKeys}`);
```

What do you think?

Reply:
As a user of EdgeDB, I'd prefer it if the errors were specific to the conditions that caused them. Unfortunately, edgedb server doesn't return enough information in the headers to determine the cause of the error, so we need to look at the response body. This is especially true in the second request, where three different issues: a bad session ID (which could be a client error or the result of server restart/request being routed to a different instance), a session timeout (client error), and invalid credentials (client misconfiguration) — all produce the same status/headers combo. In all cases though, edgedb-generated errors are succinct and don't include any sensitive data, so I think it's reasonably safe to include them into the error message.

However, I agree that there's a risk of getting an unwieldy response body here from some intermediate HTTP gateway spewing an HTML error page. We can guard against that by truncating the error message. Additionally, we could check the content type before reading the response body — it looks like edgedb returns `Content-Type: text/plain` in error responses.

Reply:
> However, I agree that there's a risk of getting an unwieldy response body here from some intermediate HTTP gateway spewing an HTML error page. We can guard against that by truncating the error message. Additionally, we could check the content type before reading the response body — it looks like edgedb returns Content-Type: text/plain in error responses.

Yeah, I don't want to get too caught up in guarding this since we're replacing a literal `console.log` here (🙈 ), but it would be good to avoid buffering the whole response to protect against some DDoS/accidental bug where the streaming response trickles in or it spews gigs of text down the stream. I haven't used `ReadableStream` yet, but I think something like this could work (export from the `utils.ts` module?):

```ts
export async function streamToFixedBuffer(
  size: number,
  stream: ReadableStream
): Promise<Buffer> {
  const buf = Buffer.alloc(size);
  let bytesRead = 0;

  for await (const chunk of stream) {
    const remainingSpace = size - bytesRead;
    buf.set(chunk.slice(0, remainingSpace), bytesRead);
    bytesRead += remainingSpace;

    if (bytesRead === size) {
      break;
    }
  }

  return buf;
}
```

[CodeSandbox](https://codesandbox.io/p/sandbox/naughty-morse-7x9eop?file=%2Findex.mjs%3A19%2C21)

Then used like:

```suggestion
    const body = await streamToFixedBuffer(1024, serverFirstRes.body);
    throw new ProtocolError(`authentication failed: ${body}`);
```

---

Apologies for falling into this rabbit hole, this is something I want to improve in general and it's helpful to have someone to bounce ideas off of here.

Reply:
Let's ship what we have here. I want to improve tracing/logging in general so we can do something more fundamentally helpful and general at the point to mitigate any issues then.

Reply:
FWIW, I was able to come up with a fairly compact version of `streamToFixedBuffer` using a `reader` (and fix a bug 🙈 ):

```ts
export async function streamToFixedBuffer(
  size: number,
  stream: ReadableStream
): Promise<Buffer> {
  const reader = stream.getReader();
  const buf = Buffer.alloc(size);
  let bytesRead = 0;

  while (bytesRead < size) {
    const { done, value } = await reader.read();
    if (done) {
      return buf;
    }

    const remainingSpace = size - bytesRead;
    const byteSlice = value.slice(0, remainingSpace);
    buf.set(byteSlice, bytesRead);
    bytesRead += byteSlice.length;

    if (bytesRead === size) {
      return buf;
    }
  }
}
```